File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
    stack<TreeNode*> nodeStack; 
    void pushLeft(TreeNode* root){
        while(root){
            nodeStack.push(root); 
            root = root->left; 
        }
    }
public:
    BSTIterator(TreeNode* root) {
        pushLeft(root); 
    }
    
    int next() {
        TreeNode* curr = nodeStack.top();
        nodeStack.pop();

        pushLeft(curr->right); 
        
        return curr->val; 
    }
    
    bool hasNext() {
        return !nodeStack.empty(); 
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {

    List<TreeNode> inOrder = new ArrayList<>(); 
    int index = 0; 

    public BSTIterator(TreeNode root) {
        helper(root,inOrder); 
    }
    
    public int next() {
        return inOrder.get(index++).val; 
    }
    
    public boolean hasNext() {
        return index < inOrder.size(); 
    }

    public void helper(TreeNode root, List<TreeNode> inOrder){
        if(root == null) return; 

        helper(root.left, inOrder); 
        inOrder.add(root); 
        helper(root.right, inOrder); 
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */